[export] Make aoti_call_delegate hop traceable

yiming0416 opened the pull request:
Summary: The `aoti_call_delegate` hop now uses a stateless `original_gm` for tracing with fake tensors and the OSS AOTI Runner for running with real tensors


Differential Revision: D70738393




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ydwu4 commented on torch/_higher_order_ops/aoti_call_delegate.py:
```diff
@@ -77,23 +76,49 @@ def call_delegate_cpu(
+    has_fake_args = any(isinstance(arg, FakeTensor) for arg in new_args)
+    if has_fake_args:
+        # use stateless original_gm for tracing with fake tensors
+        return original_gm(*new_args)
```
Move this to FakeTensorMode

ydwu4 commented on torch/_higher_order_ops/aoti_call_delegate.py:
```diff
@@ -77,23 +76,49 @@ def call_delegate_cpu(
+        "call_function", func_overload, proxy_args, {}, name="aoti_call_delegate"
     )
+    with disable_proxy_modes_tracing():
+        out = call_delegate_cpu(lowered_module, original_gm, weight_args, input_args)
```
out is tuple but downstream expect it to be tensor. Constract should be the original output signature is preserved.
1. get the original signature from original_gm
2. cat the output of out_proxy into original_gm

extending:
out_proxy[0] -> get_item_proxy
track_tensor_tree(out[0], get_item_proxy)

Maybe still need to call track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)

original_proxy = pytree.tree_unflatten(out_proxy)
original_example_value  = pytree.tree_unflatten(out)
track_tensor_tree(original_proxy, original_example_value )

yiming0416 committed:
[export] Make aoti_call_delegate hop traceable (#148804)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/148804

The `aoti_call_delegate` hop now uses a stateless `original_gm` for tracing with fake tensors and the OSS AOTI Runner for running with real tensors

Test Plan:
```
buck2 run mode/dev-nosan sigmoid/inference/test:test_partial_lowering -- -r test_partial_lowering_single_output

buck2 run mode/dev-nosan sigmoid/inference/test:test_partial_lowering -- -r test_partial_lowering_multi_outputs

buck2 run mode/dev-nosan sigmoid/inference/test:test_partial_lowering -- -r test_partial_lowering_pytree_outputs_1

buck2 run mode/dev-nosan sigmoid/inference/test:test_partial_lowering -- -r test_partial_lowering_pytree_outputs_2
```

Reviewed By: SherlockNoMad

Differential Revision: D70738393
